src/output_writers.jl

```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
         return false
     end
 end
+
+mutable struct Checkpointer <: OutputWriter
+                 dir :: String
+              prefix :: String
+    output_frequency :: Int
+end
+
+function Checkpointer(; output_frequency, dir=".", prefix="checkpoint", force=false)
+    mkpath(dir)
+    return Checkpointer(dir, prefix, output_frequency)
+end
+
+function savesubfields!(file, model, name, flds=propertynames(getproperty(model, name)))
+    for f in flds
+        file["$name/$f"] = Array(getproperty(getproperty(model, name), f).data.parent)
+    end
+    return nothing
+end
+
+checkpointed_structs   = [:arch, :boundary_conditions, :grid, :clock, :eos, :constants, :closure]
+checkpointed_fieldsets = [:velocities, :tracers, :G, :Gp]
+
+function write_output(model, c::Checkpointer)
+    @warn "Checkpointer will not save forcing functions, output writers, or diagnostics. They will need to be " *
+          "restored manually."
+
+    filepath = joinpath(c.dir, c.prefix * string(model.clock.iteration) * ".jld2")
+
+    jldopen(filepath, "w") do file
+        # Checkpointing model properties that we can just serialize.
+        [file["$p"] = getproperty(model, p) for p in checkpointed_structs]
+
+        # Checkpointing structs containing fields.
+        [savesubfields!(file, model, p) for p in checkpointed_fieldsets]
+    end
+end
+
+_arr(::CPU, a) = a
+_arr(::GPU, a) = CuArray(a)
+
+function restore_from_checkpoint(filepath)
```

Comment:
This function must have a `; kwargs...` argument which is then merged with the retrieved `kwargs` from the checkpoint file before being passed to the model constructor. This is needed to restore models with forcing functions or non-default boundary conditions.